kaldb/src/test/java/com/slack/kaldb/server/KaldbTest.java

```diff
@@ -0,0 +1,304 @@
+package com.slack.kaldb.server;
+
+import static com.slack.kaldb.logstore.LuceneIndexStoreImpl.MESSAGES_RECEIVED_COUNTER;
+import static com.slack.kaldb.server.KaldbConfig.DEFAULT_START_STOP_DURATION;
+import static com.slack.kaldb.testlib.KaldbGrpcQueryUtil.runHealthCheckOnPort;
+import static com.slack.kaldb.testlib.KaldbGrpcQueryUtil.searchUsingGrpcApi;
+import static com.slack.kaldb.testlib.MetricsUtil.getCount;
+import static com.slack.kaldb.testlib.TestKafkaServer.produceMessagesToKafka;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.awaitility.Awaitility.await;
+
+import com.adobe.testing.s3mock.junit4.S3MockRule;
+import com.slack.kaldb.chunkManager.RollOverChunkTask;
+import com.slack.kaldb.proto.config.KaldbConfigs;
+import com.slack.kaldb.proto.service.KaldbSearch;
+import com.slack.kaldb.testlib.KaldbConfigUtil;
+import com.slack.kaldb.testlib.MessageUtil;
+import com.slack.kaldb.testlib.TestKafkaServer;
+import io.micrometer.core.instrument.search.MeterNotFoundException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import org.apache.curator.test.TestingServer;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.CreateBucketRequest;
+
+public class KaldbTest {
+  private static final Logger LOG = LoggerFactory.getLogger(KaldbTest.class);
+
+  private static final String TEST_S3_BUCKET = "test-s3-bucket";
+  private static final String TEST_KAFKA_TOPIC_1 = "test-topic-1";
+  private static final String KALDB_TEST_CLIENT_1 = "kaldb-test-client1";
+  private static final String KALDB_TEST_CLIENT_2 = "kaldb-test-client2";
+
+  @ClassRule public static final S3MockRule S3_MOCK_RULE = S3MockRule.builder().silent().build();
+  private TestKafkaServer kafkaServer;
+  private TestingServer zkServer;
+  private S3Client s3Client;
+
+  @Before
+  public void setUp() throws Exception {
+    zkServer = new TestingServer(2181);
```

Comment:
We should use either randomly allocated ports or different port numbers; If you have a locally running ZK or Kafka (ie, docker container) you would need to shut it down before running this test locally.

Reply:
Currently, the `testBootAllComponentsStartSuccessfullyFromConfig` loads the config.yaml that is used by our docker container in bedrock. Since this config has defaults that talk to a local ZK cluster and kafka cluster, we need to start the unit tests on these ports. We can change the defaults in the config but I don't think that's recommended. So, for now it may be better to leave these as is until it causes an issue.

Reply:
You can do this by just setting the token values to whatever port you've assigned using a `StringSubstitutor`. This is how the yaml is constructed today - we haven't made this into a general test util or overloaded constructor yet, as we haven't had the need until now.

```java
@Test
public void testBootAllComponentsStartSuccessfullyFromConfig() throws Exception {
  Map<String, String> values = ImmutableMap.of(
    "KAFKA_BOOTSTRAP_SERVERS", "localhost:" + kafkaServer.getBroker().getKafkaPort().get(),
    "KALDB_ZK_CONNECTION_STRING", "localhost:" + zkServer.getPort()
  );
  StringSubstitutor substitute = new StringSubstitutor(s -> values.getOrDefault(s, null));

  KaldbConfigs.KaldbConfig kaldbConfig = 
  KaldbConfig.fromYamlConfig(substitute.replace(Files.readString(Path.of("../config/config.yaml"))));
```

If we think we'll want additional integration tests that can set various config values for testing purposes, we can just overload the `fromYamlConfig` to take an optional Map of key/value pairs, instead of only loading the system env map.

```java
static KaldbConfigs.KaldbConfig fromYamlConfig(String yamlStr)
static KaldbConfigs.KaldbConfig fromYamlConfig(String yamlStr, Map<String, String> values)
```

Reply:
Thanks. Updated the code to remove hard coded ports.